src/neuroglancer/ui/hiding_list.ts

```diff
@@ -0,0 +1,196 @@
+  }
+
+  private addElementHelper(element: HTMLElement) {
+    this.scrollArea.appendChild(element);
+    const h = element.offsetHeight;
+    this.elementIndices.set(element, this.elementHeights.length);
+    this.elementHeights.push([element, this.totalH]);
+    this.totalH += h;
+    this.hideElement(element);
+    this.resizeObserver.observe(element);
+  }
+
+  addElements(elements: HTMLElement[]) {
+    // append many at once for better performance
+    for (const element of elements) {
+      this.addElementHelper(element);
+    }
+    this.updateScrollHeight();
+    this.updateScrollAreaPos();
+  }
+
+  addElement(element: HTMLElement) {
+    this.addElementHelper(element);
+    this.updateScrollHeight();
+    this.updateScrollAreaPos();
+  }
+
+  removeElement(element: HTMLElement) {
+    this.resizeObserver.unobserve(element);
+    this.unhideElement(element);
+    const h = element.offsetHeight;
+
+    const i = this.elementIndices.get(element)!;
+    this.elementHeights.splice(i, 1);
+    this.elementIndices.delete(element);
+    this.shiftHeightsAfter(i, -h);
+    // shift indices of elements that came after the removed one
+    for (let j = i; j < this.elementHeights.length; j++) {
+      const el = this.elementHeights[j][0];
+      this.elementIndices.set(el, j);
+    }
+
+    this.totalH -= h;
+    this.scrollArea.removeChild(element);
+    this.updateScrollHeight();
+    this.updateScrollAreaPos();
+  }
+
+  removeAll() {
+    for (const [element] of this.elementIndices) {
+      this.scrollArea.removeChild(element);
+    }
+
+    this.elementHeights = [];
+    this.elementIndices = new Map<HTMLElement, number>();
+
+    this.totalH = 0;
+    this.loadedElements = [];
+  }
+
+  scrollTo(element: HTMLElement) {
+    const h = this.elementHeights[this.elementIndices.get(element)!][1];
+    // Scrolls just a pixel too far, this makes it look prettier
+    this.scrollbar.scrollTop = h - 1;
+    this.updateScrollAreaPos();
+  }
+
+  recalculateHeights() {
+    this.totalH = 0;
+    this.scrollbar.scrollTop = 0;
+    for (const [element, i] of this.elementIndices) {
+      this.unhideElement(element);
+      const h = element.offsetHeight;
+      this.elementHeights[i][1] = this.totalH;
+      this.totalH += h;
+    }
+    this.updateScrollHeight();
+    this.updateScrollAreaPos();
+  }
+
+  private hideElement(element: HTMLElement) {
+    element.classList.add('neuroglancer-annotation-hiding-list-hiddenitem');
+  }
+
+  private unhideElement(element: HTMLElement) {
+    element.classList.remove('neuroglancer-annotation-hiding-list-hiddenitem');
+  }
+
+  private updateScrollHeight() {
+    // Add some extra padding on the bottom
+    this.scrollbarFiller.style.height = (this.totalH + 10) + 'px';
+  }
+
+  private shiftHeightsAfter(startIndex: number, delta: number) {
+    for (let i = startIndex; i < this.elementHeights.length; i++) {
+      this.elementHeights[i][1] += delta;
+    }
+  }
+
+  private findHeight(h: number) {
```

Comment:
A comment explaining why we need to find this specific height, and mentioning that this function performs a binary search, would be quite helpful. I also think the variable names could be a little more descriptive (l -> left, r -> right, m -> mid).